pkg/services/datasources/service/datasource.go

```diff
@@ -534,21 +534,26 @@ func (s *Service) UpdateDataSource(ctx context.Context, cmd *datasources.UpdateD
 			}
 		}
 
-		// TODO: we will eventually remove this check for moving the resource to it's separate API
-		if s.features != nil && s.features.IsEnabled(ctx, featuremgmt.FlagTeamHttpHeaders) && !cmd.OnlyUpdateLBACRulesFromAPI {
+		// TODO: Refactor to store lbac rules separate from a datasource
+		updatingLBACRules := cmd.OnlyUpdateLBACRulesFromAPI && s.features != nil && s.features.IsEnabled(ctx, featuremgmt.FlagTeamHttpHeaders) && cmd.JsonData != nil && cmd.JsonData.Get("teamHttpHeaders").Interface() != nil
+		if updatingLBACRules {
 			s.logger.Debug("Overriding LBAC rules with stored ones",
 				"reason", "update_lbac_rules_from_datasource_api",
 				"action", "use_updateLBACRules_API",
 				"datasource_id", dataSource.ID,
 				"datasource_uid", dataSource.UID)
-
+			cmd.JsonData.Set("teamHttpHeaders", cmd.JsonData.Get("teamHttpHeaders").Interface())
+		} else {
+			// make sure to preserve existing lbac rules when updating datasource, regardless of if we're updating lbac rules or not
+			var existingTeamHttpHeaders interface{}
```

Comment:
I preferred the previous code because it was more easier to understand.
I think we are introducing a regression, we are not accounting for the following case anymore: "Should remove the rules if the data-source had no rule before."

Reply:
addressing in https://github.com/grafana/grafana/pull/94622#discussion_r1808444530